File: src/vs2026/Services/PipelineYamlAnalyzer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using PipelinesExplorer.VisualStudio.AzureDevOps;
using YamlDotNet.Core;
using YamlDotNet.RepresentationModel;

namespace PipelinesExplorer.VisualStudio.Services;

/// <summary>Reference to a YAML template (`template:` key).</summary>
public sealed class TemplateRef
{
    /// <summary>Raw <c>template:</c> value as written in YAML.</summary>
    public string Raw { get; }
    /// <summary>Local path portion (no <c>@repo</c> suffix).</summary>
    public string Path { get; }
    /// <summary>Optional repository alias if the template lives in another repo.</summary>
    public string? Repository { get; }

    public TemplateRef(string raw, string path, string? repository)
    {
        Raw = raw;
        Path = path;
        Repository = repository;
    }
}

/// <summary>Detected script flavour for a <see cref="ScriptRef"/>.</summary>
public enum ScriptKind
{
    PowerShell,
    Bash,
    Cmd,
    Python,
    AzureCli,
    Unknown,
}

/// <summary>Reference to a script-style task (or shorthand step) in a pipeline.</summary>
public sealed class ScriptRef
{
    public string Task { get; }
    public string? FilePath { get; }
    public bool Inline { get; }
    public int? Line { get; }
    public ScriptKind Kind { get; }

    public ScriptRef(string task, string? filePath, bool inline, int? line, ScriptKind kind)
    {
        Task = task;
        FilePath = filePath;
        Inline = inline;
        Line = line;
        Kind = kind;
    }
}

/// <summary>Result of analysing a single YAML file.</summary>
public sealed class PipelineAnalysis
{
    public IReadOnlyList<TemplateRef> Templates { get; init; } = Array.Empty<TemplateRef>();
    public IReadOnlyList<ScriptRef> Scripts { get; init; } = Array.Empty<ScriptRef>();
    public string? RootPath { get; init; }
    public string? Warning { get; init; }
}

/// <summary>
/// Loads YAML for pipelines or individual templates and walks the AST to
/// extract referenced templates and script-running tasks. Mirrors the
/// VS Code client's <c>PipelineYamlAnalyzer</c>.
/// </summary>
public sealed partial class PipelineYamlAnalyzer
{
    [GeneratedRegex(@"^([A-Za-z][A-Za-z0-9]*)@\d+$")]
    private static partial Regex TaskNameRe();

    /// <summary>Map of fully-qualified ADO task names (lowercased, no <c>@N</c>) to their default kind.</summary>
    private static readonly Dictionary<string, ScriptKind> TaskKindMap = new(StringComparer.OrdinalIgnoreCase)
    {
        ["powershell"] = ScriptKind.PowerShell,
        ["azurepowershell"] = ScriptKind.PowerShell,
        ["powershellontargetmachines"] = ScriptKind.PowerShell,
        ["bash"] = ScriptKind.Bash,
        ["shellscript"] = ScriptKind.Bash,
        ["cmdline"] = ScriptKind.Cmd,
        ["batchscript"] = ScriptKind.Cmd,
        ["pythonscript"] = ScriptKind.Python,
        ["azurecli"] = ScriptKind.AzureCli,
    };

    /// <summary>Shorthand step keys that execute a script (alternative to <c>task:</c>).</summary>
    private static readonly Dictionary<string, ScriptKind> ShorthandKindMap = new(StringComparer.OrdinalIgnoreCase)
    {
        ["script"] = ScriptKind.Cmd,         // ADO maps `script:` to cmd on Windows / bash elsewhere
        ["bash"] = ScriptKind.Bash,
        ["pwsh"] = ScriptKind.PowerShell,
        ["powershell"] = ScriptKind.PowerShell,
    };

    private readonly AdoClient _client;
    private readonly LoggingService _logger;

    public PipelineYamlAnalyzer(AdoClient client, LoggingService logger)
    {
        _client = client;
        _logger = logger;
    }

    /// <summary>Analyse an arbitrary YAML file inside a TfsGit repository.</summary>
    public async Task<PipelineAnalysis> AnalyzeFileAsync(
        string organizationName,
        string projectName,
        string repositoryId,
        string filePath,
        string? branch = null,
        CancellationToken cancellationToken = default)
    {
        string? yamlText;
        try
        {
            yamlText = await _client.GetFileContentAsync(organizationName, projectName, repositoryId, filePath, branch, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            _logger.Error($"Failed to fetch YAML {filePath}", ex);
            return new PipelineAnalysis { RootPath = filePath, Warning = "Could not fetch YAML file." };
        }
        if (string.IsNullOrEmpty(yamlText))
        {
            return new PipelineAnalysis { RootPath = filePath, Warning = "YAML file not found in the repository." };
        }

        try
        {
            var templates = new List<TemplateRef>();
            var scripts = new List<ScriptRef>();
            using var reader = new System.IO.StringReader(yamlText!);
            var stream = new YamlStream();
            stream.Load(reader);
            foreach (var doc in stream.Documents)
            {
                Walk(doc.RootNode, templates, scripts);
            }
            return new PipelineAnalysis
            {
                Templates = DedupeTemplates(templates),
                Scripts = DedupeScripts(scripts),
                RootPath = filePath,
            };
        }
        catch (Exception ex)
        {
            _logger.Error($"Failed to parse YAML {filePath}", ex);
            return new PipelineAnalysis { RootPath = filePath, Warning = "YAML parse error." };
        }
    }

    /// <summary>Analyse a pipeline by id, fetching its definition first if needed.</summary>
    public async Task<PipelineAnalysis> AnalyzeAsync(
        string organizationName,
        string projectName,
        int pipelineId,
        AdoPipelineDetail? preloadedDetail = null,
        string? branch = null,
        CancellationToken cancellationToken = default)
    {
        AdoPipelineDetail detail;
        if (preloadedDetail is not null)
        {
            detail = preloadedDetail;
        }
        else
        {
            try
            {
                detail = await _client.GetPipelineAsync(organizationName, projectName, pipelineId, cancellationToken).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                _logger.Error($"Failed to fetch pipeline {pipelineId} definition", ex);
                return new PipelineAnalysis { Warning = "Could not fetch pipeline definition." };
            }
        }

        var cfg = detail.Configuration;
        if (cfg is null || (!string.IsNullOrEmpty(cfg.Type) && !string.Equals(cfg.Type, "yaml", StringComparison.OrdinalIgnoreCase)))
        {
            return new PipelineAnalysis { Warning = "Pipeline is not YAML-based." };
        }
        var repoId = cfg.Repository?.Id;
        var yamlPath = cfg.Path;
        if (string.IsNullOrEmpty(repoId) || string.IsNullOrEmpty(yamlPath))
        {
            return new PipelineAnalysis { Warning = "Pipeline definition has no repository / path." };
        }
        if (!string.IsNullOrEmpty(cfg.Repository?.Type)
            && !string.Equals(cfg.Repository!.Type, "azureReposGit", StringComparison.OrdinalIgnoreCase))
        {
            return new PipelineAnalysis
            {
                RootPath = yamlPath,
                Warning = $"Repository type \"{cfg.Repository.Type}\" is not supported yet.",
            };
        }

        return await AnalyzeFileAsync(organizationName, projectName, repoId!, yamlPath!, branch, cancellationToken).ConfigureAwait(false);
    }

    private static void Walk(YamlNode node, List<TemplateRef> templates, List<ScriptRef> scripts)
    {
        switch (node)
        {
            case YamlMappingNode map:
                {
                    if (TryGetScalar(map, "template", out var tplVal))
                    {
                        templates.Add(ParseTemplateRef(tplVal!));
                    }
                    if (TryGetScalar(map, "task", out var taskVal) && TryKindForTask(taskVal!, out var taskKind))
                    {
                        YamlNode? inputsNode = null;
                        foreach (var kv in map.Children)
                        {
                            if (kv.Key is YamlScalarNode s && string.Equals(s.Value, "inputs", StringComparison.Ordinal))
                            {
                                inputsNode = kv.Value;
                                break;
                            }
                        }
                        var line = map.Start.Line > 0 ? (int?)map.Start.Line : null;
                        scripts.Add(ParseTaskRef(taskVal!, taskKind, inputsNode, line));
                    }
                    else if (TryFindShorthand(map, out var shorthandKey, out var shorthandKind))
                    {
                        var line = map.Start.Line > 0 ? (int?)map.Start.Line : null;
                        scripts.Add(new ScriptRef(shorthandKey, null, true, line, shorthandKind));
                    }
                    foreach (var kv in map.Children)
                    {
                        Walk(kv.Value, templates, scripts);
                    }
                    break;
                }
            case YamlSequenceNode seq:
                foreach (var child in seq.Children) { Walk(child, templates, scripts); }
                break;
        }
    }

    /// <summary>Try to resolve a fully-qualified ADO task identifier to a <see cref="ScriptKind"/>.</summary>
    private static bool TryKindForTask(string task, out ScriptKind kind)
    {
        var m = TaskNameRe().Match(task);
        if (m.Success && TaskKindMap.TryGetValue(m.Groups[1].Value, out kind))
        {
            return true;
        }
        kind = ScriptKind.Unknown;
        return false;
    }

    /// <summary>
    /// Detect whether a step mapping uses the shorthand form (e.g. <c>- bash:</c>,
    /// <c>- pwsh:</c>, <c>- script:</c>) instead of <c>- task: Foo@N</c>. Steps
    /// that have a <c>task</c> or <c>template</c> key are excluded.
    /// </summary>
    private static bool TryFindShorthand(YamlMappingNode map, out string key, out ScriptKind kind)
    {
        string? found = null;
        ScriptKind foundKind = ScriptKind.Unknown;
        foreach (var kv in map.Children)
        {
            if (kv.Key is not YamlScalarNode s || s.Value is null) { continue; }
            if (string.Equals(s.Value, "task", StringComparison.Ordinal)
                || string.Equals(s.Value, "template", StringComparison.Ordinal))
            {
                key = string.Empty; kind = ScriptKind.Unknown;
                return false;
            }
            if (found is null && ShorthandKindMap.TryGetValue(s.Value, out var k))
            {
                found = s.Value;
                foundKind = k;
            }
        }
        if (found is null) { key = string.Empty; kind = ScriptKind.Unknown; return false; }
        key = found; kind = foundKind;
        return true;
    }

    private static bool TryGetScalar(YamlMappingNode map, string key, out string? value)
    {
        foreach (var kv in map.Children)
        {
            if (kv.Key is YamlScalarNode s && string.Equals(s.Value, key, StringComparison.Ordinal)
                && kv.Value is YamlScalarNode v)
            {
                value = v.Value;
                return value is not null;
            }
        }
        value = null;
        return false;
    }

    private static TemplateRef ParseTemplateRef(string raw)
    {
        var at = raw.LastIndexOf('@');
        if (at < 0) { return new TemplateRef(raw, raw, null); }
        return new TemplateRef(raw, raw.Substring(0, at), raw.Substring(at + 1));
    }

    private static ScriptRef ParseTaskRef(string task, ScriptKind defaultKind, YamlNode? inputs, int? line)
    {
        if (inputs is not YamlMappingNode inputsMap)
        {
            return new ScriptRef(task, null, true, line, defaultKind);
        }
        var lower = new Dictionary<string, string?>(StringComparer.OrdinalIgnoreCase);
        foreach (var kv in inputsMap.Children)
        {
            if (kv.Key is YamlScalarNode k && k.Value is not null)
            {
                lower[k.Value] = (kv.Value as YamlScalarNode)?.Value;
            }
        }

        // File-path style inputs across the supported tasks:
        //   PowerShell@2 / Bash@3 / PythonScript@0 → filePath / scriptPath
        //   AzurePowerShell@5 → ScriptPath
        //   ShellScript@2 / AzureCLI@2 → scriptPath
        //   BatchScript@1 → filename
        //   PowerShellOnTargetMachines@3 → ScriptPath
        lower.TryGetValue("filePath", out var fp1);
        lower.TryGetValue("scriptPath", out var fp2);
        lower.TryGetValue("filename", out var fp3);
        var filePath = !string.IsNullOrEmpty(fp1) ? fp1
            : !string.IsNullOrEmpty(fp2) ? fp2
            : !string.IsNullOrEmpty(fp3) ? fp3 : null;

        // Refine kind for AzureCLI based on its `scriptType` (ps / pscore / bash / batch).
        var kind = defaultKind;
        if (defaultKind == ScriptKind.AzureCli)
        {
            lower.TryGetValue("scriptType", out var st);
            var stLower = (st ?? string.Empty).ToLowerInvariant();
            if (stLower == "ps" || stLower == "pscore") { kind = ScriptKind.PowerShell; }
            else if (stLower == "bash") { kind = ScriptKind.Bash; }
            else if (stLower == "batch") { kind = ScriptKind.Cmd; }
            // otherwise keep the generic AzureCli icon.
        }

        if (!string.IsNullOrEmpty(filePath))
        {
            return new ScriptRef(task, filePath, false, line, kind);
        }
        lower.TryGetValue("targetType", out var t1);
        lower.TryGetValue("scriptType", out var t2);
        lower.TryGetValue("scriptLocation", out var t3);
        lower.TryGetValue("scriptSource", out var t4);
        var targetType = (t1 ?? t2 ?? t3 ?? t4 ?? string.Empty).ToLowerInvariant();
        var isInline =
            targetType == "inline" || targetType == "inlinescript"
            || lower.ContainsKey("script") || lower.ContainsKey("inline") || lower.ContainsKey("inlineScript");
        return new ScriptRef(task, null, isInline, line, kind);
    }

    private static IReadOnlyList<TemplateRef> DedupeTemplates(IEnumerable<TemplateRef> items)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        return items.Where(t => seen.Add(t.Raw)).ToList();
    }

    private static IReadOnlyList<ScriptRef> DedupeScripts(IEnumerable<ScriptRef> items)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        return items.Where(s =>
        {
            var key = $"{s.Task}|{s.Kind}|{s.FilePath ?? (s.Inline ? $"<inline:{s.Line?.ToString() ?? "?"}>" : "<unknown>")}";
            return seen.Add(key);
        }).ToList();
    }
}

File: src/vscode/src/pipelineYamlAnalyzer.ts
import { isMap, LineCounter, parseDocument, visit, YAMLMap } from 'yaml';
import * as vscode from 'vscode';
import { AdoClient, AdoPipelineDetail } from './adoClient';
import { LoggingService } from './LoggingService';

export interface TemplateRef {
	/** Raw `template:` value as written in YAML (e.g. `steps/build.yml@templates`). */
	raw: string;
	/** Local path portion (no `@repo` suffix). */
	path: string;
	/** Optional repository alias if the template lives in another repo. */
	repository?: string;
}

/** Detected script flavour, used to pick the icon shown in the tree. */
export type ScriptKind = 'powershell' | 'bash' | 'cmd' | 'python' | 'azurecli' | 'unknown';

export interface ScriptRef {
	/** Task identifier (e.g. `PowerShell@2`) or shorthand keyword (`script`, `bash`, `pwsh`, `powershell`). */
	task: string;
	/** Path of the script file when the task references a file. */
	filePath?: string;
	/** True if the task uses an inline script (no external file). */
	inline: boolean;
	/** 1-based line number of the task in the source YAML (when known). */
	line?: number;
	/** Detected script kind, used to choose the icon. */
	kind: ScriptKind;
}

export interface PipelineAnalysis {
	templates: TemplateRef[];
	scripts: ScriptRef[];
	/** Path of the root pipeline YAML, useful for tooltip. */
	rootPath?: string;
	/** Non-fatal warning produced while loading or parsing the YAML. */
	warning?: string;
}

/** Map of fully-qualified ADO task ids (lowercased, no `@N` suffix) to their default kind. */
const TASK_KIND_MAP: Record<string, ScriptKind> = {
	'powershell': 'powershell',
	'azurepowershell': 'powershell',
	'powershellontargetmachines': 'powershell',
	'bash': 'bash',
	'shellscript': 'bash',
	'cmdline': 'cmd',
	'batchscript': 'cmd',
	'pythonscript': 'python',
	'azurecli': 'azurecli',
};

/** Shorthand step keys (alternative to `task:`) that execute a script. */
const SHORTHAND_KIND_MAP: Record<string, ScriptKind> = {
	'script': 'cmd',          // ADO maps `script:` to cmd on Windows / bash elsewhere
	'bash': 'bash',
	'pwsh': 'powershell',
	'powershell': 'powershell',
};

const TASK_NAME_RE = /^([A-Za-z][A-Za-z0-9]*)@\d+$/;

/**
 * Loads YAML for pipelines or individual templates and walks the AST to extract
 * referenced templates and PowerShell-style tasks. Same-repo templates can be
 * recursed into via {@link analyzeFile}.
 */
export class PipelineYamlAnalyzer {
	constructor(
		private readonly client: AdoClient,
		private readonly logger: LoggingService,
	) { }

	/**
	 * Analyse an arbitrary YAML file inside a TfsGit repository. Used to recurse
	 * into templates referenced from the root pipeline (or another template).
	 */
	async analyzeFile(
		organizationName: string,
		projectName: string,
		repositoryId: string,
		filePath: string,
		branch?: string,
	): Promise<PipelineAnalysis> {
		let yamlText: string | undefined;
		try {
			yamlText = await this.client.getFileContent(organizationName, projectName, repositoryId, filePath, branch);
		} catch (err) {
			this.logger.logError(`Failed to fetch YAML ${filePath}`, err);
			return { templates: [], scripts: [], rootPath: filePath, warning: vscode.l10n.t('Could not fetch YAML file.') };
		}
		if (!yamlText) {
			return { templates: [], scripts: [], rootPath: filePath, warning: vscode.l10n.t('YAML file not found in the repository.') };
		}
		try {
			const lineCounter = new LineCounter();
			const doc = parseDocument(yamlText, { keepSourceTokens: false, lineCounter });
			const templates: TemplateRef[] = [];
			const scripts: ScriptRef[] = [];
			walkAst(doc, lineCounter, templates, scripts);
			return { templates: dedupeTemplates(templates), scripts: dedupeScripts(scripts), rootPath: filePath };
		} catch (err) {
			this.logger.logError(`Failed to parse YAML ${filePath}`, err);
			return { templates: [], scripts: [], rootPath: filePath, warning: vscode.l10n.t('YAML parse error.') };
		}
	}

	async analyze(
		organizationName: string,
		projectName: string,
		pipelineId: number,
		preloadedDetail?: AdoPipelineDetail,
		branch?: string,
	): Promise<PipelineAnalysis> {
		let detail: AdoPipelineDetail;
		if (preloadedDetail) {
			detail = preloadedDetail;
		} else {
			try {
				detail = await this.client.getPipeline(organizationName, projectName, pipelineId);
			} catch (err) {
				this.logger.logError(`Failed to fetch pipeline ${pipelineId} definition`, err);
				return { templates: [], scripts: [], warning: vscode.l10n.t('Could not fetch pipeline definition.') };
			}
		}

		const cfg = detail.configuration;
		if (!cfg || (cfg.type && cfg.type.toLowerCase() !== 'yaml')) {
			return { templates: [], scripts: [], warning: vscode.l10n.t('Pipeline is not YAML-based.') };
		}
		const repoId = cfg.repository?.id;
		const yamlPath = cfg.path;
		if (!repoId || !yamlPath) {
			return { templates: [], scripts: [], warning: vscode.l10n.t('Pipeline definition has no repository / path.') };
		}
		if (cfg.repository?.type && cfg.repository.type.toLowerCase() !== 'azurereposgit') {
			return {
				templates: [], scripts: [], rootPath: yamlPath,
				warning: `Repository type "${cfg.repository.type}" is not supported yet.`,
			};
		}

		return this.analyzeFile(organizationName, projectName, repoId, yamlPath, branch);
	}
}

/**
 * Walk the YAML AST (rather than the JS object) so we can capture source
 * line numbers for every script-style task. The `visit` traversal naturally
 * descends into `extends:` so its inner `template:` pair is picked up by
 * the same Map handler. In addition to the canonical `- task: Foo@N` form
 * we also recognise the shorthand step keys (`script`, `bash`, `pwsh`,
 * `powershell`) that ADO accepts as an alternative.
 */
function walkAst(
	doc: ReturnType<typeof parseDocument>,
	lineCounter: LineCounter,
	templates: TemplateRef[],
	scripts: ScriptRef[],
): void {
	visit(doc, {
		Map(_key, node) {
			const tplVal = node.get('template');
			if (typeof tplVal === 'string') {
				templates.push(parseTemplateRef(tplVal));
			}

			const taskVal = node.get('task');
			if (typeof taskVal === 'string') {
				const taskKind = kindForTask(taskVal);
				if (taskKind !== undefined) {
					const inputsNode = node.get('inputs');
					const inputsJs = isMap(inputsNode)
						? (inputsNode as YAMLMap).toJSON()
						: inputsNode;
					const line = lineOf(node, lineCounter);
					scripts.push(parseTaskRef(taskVal, taskKind, inputsJs, line));
					return;
				}
			}

			// Shorthand step keys: a step is a map whose first/only meaningful key is
			// `script` / `bash` / `pwsh` / `powershell`. Treat them as inline scripts.
			const shorthand = findShorthandKey(node);
			if (shorthand) {
				const line = lineOf(node, lineCounter);
				scripts.push({
					task: shorthand.key,
					inline: true,
					line,
					kind: SHORTHAND_KIND_MAP[shorthand.key],
				});
			}
		},
	});
}

function lineOf(node: { range?: [number, number, number] | null }, lineCounter: LineCounter): number | undefined {
	const range = node.range;
	return range ? lineCounter.linePos(range[0]).line : undefined;
}

/**
 * Find the shorthand script key (if any) on a step mapping. The key must be
 * one of {@link SHORTHAND_KIND_MAP}; we only treat it as a script step if a
 * `task:` key is *not* present (those are handled separately above).
 */
function findShorthandKey(node: YAMLMap): { key: keyof typeof SHORTHAND_KIND_MAP } | undefined {
	let found: keyof typeof SHORTHAND_KIND_MAP | undefined;
	for (const item of node.items) {
		const k = (item.key as { value?: unknown } | null)?.value;
		if (typeof k !== 'string') { continue; }
		if (k === 'task' || k === 'template') { return undefined; }
		const lower = k.toLowerCase();
		if (lower in SHORTHAND_KIND_MAP && found === undefined) {
			found = lower as keyof typeof SHORTHAND_KIND_MAP;
		}
	}
	return found ? { key: found } : undefined;
}

/**
 * Resolve the {@link ScriptKind} for a fully-qualified task identifier
 * (e.g. `PowerShell@2`). Returns `undefined` if the task is not a known
 * script-running task.
 */
function kindForTask(task: string): ScriptKind | undefined {
	const m = TASK_NAME_RE.exec(task);
	if (!m) { return undefined; }
	const lower = m[1].toLowerCase();
	return TASK_KIND_MAP[lower];
}

function parseTemplateRef(raw: string): TemplateRef {
	const at = raw.lastIndexOf('@');
	if (at === -1) {
		return { raw, path: raw };
	}
	return { raw, path: raw.slice(0, at), repository: raw.slice(at + 1) };
}

function parseTaskRef(task: string, defaultKind: ScriptKind, inputs: unknown, line?: number): ScriptRef {
	if (!inputs || typeof inputs !== 'object') {
		return { task, inline: true, line, kind: defaultKind };
	}
	const i = inputs as Record<string, unknown>;
	// Normalise input keys (YAML is case-insensitive in practice for ADO tasks).
	const lower: Record<string, unknown> = {};
	for (const k of Object.keys(i)) {
		lower[k.toLowerCase()] = i[k];
	}

	// File-path style inputs across the supported tasks:
	//   PowerShell@2 / Bash@3 / PythonScript@0 → filePath / scriptPath
	//   AzurePowerShell@5 → ScriptPath
	//   ShellScript@2 / AzureCLI@2 → scriptPath
	//   BatchScript@1 → filename
	//   PowerShellOnTargetMachines@3 → ScriptPath
	const filePath =
		(typeof lower['filepath'] === 'string' && (lower['filepath'] as string)) ||
		(typeof lower['scriptpath'] === 'string' && (lower['scriptpath'] as string)) ||
		(typeof lower['filename'] === 'string' && (lower['filename'] as string)) ||
		undefined;

	// Refine kind for AzureCLI based on its `scriptType` (ps / pscore / bash / batch).
	let kind = defaultKind;
	if (defaultKind === 'azurecli') {
		const st = String(lower['scripttype'] ?? '').toLowerCase();
		if (st === 'ps' || st === 'pscore') { kind = 'powershell'; }
		else if (st === 'bash') { kind = 'bash'; }
		else if (st === 'batch') { kind = 'cmd'; }
		// otherwise keep the generic `azurecli` icon.
	}

	if (filePath) {
		return { task, filePath, inline: false, line, kind };
	}
	const targetType = String(
		lower['targettype'] ?? lower['scripttype'] ?? lower['scriptlocation'] ?? lower['scriptsource'] ?? '',
	).toLowerCase();
	const isInline =
		targetType === 'inline' ||
		targetType === 'inlinescript' ||
		!!lower['script'] ||
		!!lower['inline'] ||
		!!lower['inlinescript'];
	return { task, inline: isInline, line, kind };
}

function dedupeTemplates(items: TemplateRef[]): TemplateRef[] {
	const seen = new Set<string>();
	return items.filter(t => {
		if (seen.has(t.raw)) {
			return false;
		}
		seen.add(t.raw);
		return true;
	});
}

function dedupeScripts(items: ScriptRef[]): ScriptRef[] {
	const seen = new Set<string>();
	return items.filter(s => {
		const key = `${s.task}|${s.kind}|${s.filePath ?? (s.inline ? `<inline:${s.line ?? '?'}>` : '<unknown>')}`;
		if (seen.has(key)) {
			return false;
		}
		seen.add(key);
		return true;
	});
}
